src/main/java/com/pinterest/secor/uploader/HadoopS3UploadManager.java

```diff
@@ -60,12 +61,33 @@ public HadoopS3UploadManager(SecorConfig config) {
         LOG.info("uploading file {} to {}", localLogFilename, logFileName);
 
         final Future<?> f = executor.submit(new Runnable() {
+            final int totalRetries = Integer.parseInt(mConfig.getUploaderRetries());
+            final long backoffMillis = Long.parseLong(mConfig.getUploaderRetryBackoffMillis());
             @Override
             public void run() {
-                try {
-                    FileUtil.moveToCloud(localLogFilename, logFileName);
-                } catch (IOException e) {
-                    throw new RuntimeException(e);
+                int retriesLeft = totalRetries;
+                boolean success = false;
+                while (retriesLeft > 0 && !success) {
+                    try {
+                        FileUtil.moveToCloud(localLogFilename, logFileName);
+                        success = true;
+                    } catch (IOException e) {
+                        success = false;
+                        retriesLeft--;
+                        if (retriesLeft < 0)
+                            throw new RuntimeException(e);
+                        else { 
+                            LOG.warn("Error uploading files to s3. Will retry {} times. Exception {}", retriesLeft, e);
+                            try {
+                                long millisToSleep = (long)(backoffMillis*Math.pow(2, totalRetries-retriesLeft-1));
```

Comment:
This math is a bit awkward to read.  I think you can either do 'milliToSleep *= 2' or use a more forward looping variable, e.g.:
    for (int i = 0; i < numRetries; i++) {
        ...
        long millisToSleep = backoffMillis * Math.pow(2, i);
    }